### PR TITLE
[Android] Fix font weight numeric values

### PR DESCRIPTION
**Pull request by fabriziobertoglio1987**
## Summary

This issue fixes https://github.com/facebook/react-native/issues/25696 fixes https://github.com/facebook/react-native/issues/28854 fixes https://github.com/facebook/react-native/issues/26193
Since Android API 28 it is possible to specify fontWeight with numerical values ranging from 100 to 900

This pr uses the new Typeface.create() method available on Android API 28+ to set font weight value ranging from 100 to 900, while still keeping existing functionalities (custom fonts, bold/italic and other styles).
https://developer.android.com/reference/android/graphics/Typeface#create(android.graphics.Typeface,%20int,%20boolean)

## Changelog

[Android] [Fixed] - Fix font weight numeric values

## Test Plan

Works in all scenarios.

**<details><summary>CLICK TO OPEN TESTS RESULTS</summary>**
<p>

| **BEFORE** | **AFTER** | 
|:-------------------------:|:-------------------------:|
|  |  |

| **AFTER** | **AFTER** | 
|:-------------------------:|:-------------------------:|
|  |  |

| **AFTER** |
|:-------------------------:|
|  | 

</p>
</details>

### CODE REVIEW

**Comment by fangasvsass** [comment]
Looks good!

**Comment by fabriziobertoglio1987** [comment]
The microsoft tests are broken

1) I did not change the js logic
2) The changes have nothing to do with Pressability 
3) The changes have nothing to do with Microsoft platform

Unluckily the biggest problem of tests is that they take you lot of time to write, mantain and sometimes those expectations do not really troubleshoot software issues. If you tried to run the react-native test suite on local environment, you'll end up troubleshooting infinite amout of errors see https://github.com/facebook/react-native/issues/28870

I would be happy to work on the Android test suite and make it easier to run on local env, but somebody should open and assign me some issues that  I can solve on the Android Test Environment. Thanks

```
FAIL Libraries/Pressability/__tests__/Pressability-test.js (5.915 s)
  ● Pressability › onPressOut › is called after the minimum press duration by default

    expect(jest.fn()).not.toBeCalled()

    Expected number of calls: 0
    Received number of calls: 1

    1: {"bubbles": null, "cancelable": null, "currentTarget": 42, "defaultPrevented": null, "dispatchConfig": {"registrationName": "onResponderRelease"}, "eventPhase": null, "isDefaultPrevented": [Function mockConstructor], "isPropagationStopped": [Function mockConstructor], "isTrusted": null, "nativeEvent": {"changedTouches": [[Circular]], "force": 1, "identifier": 42, "locationX": 0, "locationY": 0, "pageX": 0, "pageY": 0, "target": 42, "timestamp": 1075881600000, "touches": [[Circular]]}, "persist": [Function mockConstructor], "preventDefault": [Function mockConstructor], "stopPropagation": [Function mockConstructor], "target": null, "timeStamp": 1075881600000, "touchHistory": {"indexOfSingleActiveTouch": 0, "mostRecentTimeStamp": 1075881600000, "numberActiveTouches": 1, "touchBank": []}, "type": null}

      654 | 
      655 |       jest.advanceTimersByTime(120);
    > 656 |       expect(config.onPressOut).not.toBeCalled();
          |                                     ^
      657 |       jest.advanceTimersByTime(10);
      658 |       expect(config.onPressOut).toBeCalled();
      659 |     });

      at Object.<anonymous> (Libraries/Pressability/__tests__/Pressability-test.js:656:37)


Test Suites: 1 failed, 127 passed, 128 total
```

**Comment by JoshuaGross** [comment]
@fabriziobertoglio1987 What about iOS?

**Comment by fabriziobertoglio1987** [comment]
@JoshuaGross The below are tests on iOS emulator from latest master branch (**BEFORE**) and from fix-font-weight branch (**AFTER**). I don't detect any issue on iOS. 

Additionally I added a new commit https://github.com/facebook/react-native/pull/29117/commits/e03864e72e8a95aa6f48f3e182186c1c5f48ce00 to include same example on ios RNTester.
Please contact me for any further work, improvements or issues. I will immediately fix them.

Thanks a lot
I wish you a good day
Fabrizio Bertoglio

| **BEFORE** | **AFTER** | 
|:-------------------------:|:-------------------------:|
|  |  |

| **AFTER** https://github.com/facebook/react-native/pull/29117/commits/e03864e72e8a95aa6f48f3e182186c1c5f48ce00 | 
|:-------------------------:|
|  |

| **AFTER** | **AFTER** | 
|:-------------------------:|:-------------------------:|
|  |  |

**Comment by LeeKyungJoon** [comment]
@fabriziobertoglio1987 
Hey, it says your Merge has failed. As a result, fontWeight in android is still not applied to numeral values. Can you check it again?

**Comment by fabriziobertoglio1987** [comment]
Thanks for reviewing this pull request @LeeKyungJoon, currently I am working on another Pull Request full-time and I'm sending you a quick reply. I will come back to this pull request and re-execute all the tests/verify everything is working correctly in 2-3 days.

>Hey, it says your Merge has failed. As a result, fontWeight in android is still not applied to numeral values. Can you check it again?

Your message is not not clear to me. I don't know if you are trying to implement this fix/pull request, in such case you will need to [build ReactAndroid from source](https://github.com/facebook/react-native/wiki/Building-from-source#android). 

This pull request is not yet merged to master. Once the pull request is merged in the react-native master branch, you can expect it to be available as npm package as a release candidate for testing, then will be published as release. You can upgrade at that time react-native dependency and use the upgrade tool and use the numeric falues for the font. You can expect it to be available as npm package in 2-6 months.

Instead, If you are referring to potential merge conflicts from this commit https://github.com/facebook/react-native/pull/29117/commits/e03864e72e8a95aa6f48f3e182186c1c5f48ce00 and to the fact this branch was written on an Ubuntu Machine and then tested on MacBook machine. I did not make any changes to this pull request using the MacBook, but just checkout and test this branch. The commits were all written on my Ubuntu machine, as sometimes using 2 different machines to change the same branch will create merge conflicts. I rebased before publishing the Pull Request and all CI tests are passing.

Thanks a lot
I remain available for further help
Fabrizio Bertoglio

**Comment by LeeKyungJoon** [comment]
@fabriziobertoglio1987 
답변해줘서 고마워. 지금 당장으로썬 필요하지는 않아. 다만, 조금 더 정확한 fontWeight를 구현하고 싶어서 numeral values 사용하려고 했어. 지금 react-native 63.0을 사용 중인데 혹시 이 version에 적용이 되어야하는데 적용이 안되는건가 했었어. 그리고 내 질문 중에 merge가 실패했다는 말은 Merging is Blocked라는 표현을 오해한거야. 미안해.
"You can expect it to be available as npm package in 2-6 months."
이 말을 믿고 기다릴게.
다시 한번 답변해줘서 고마워

**Comment by stackia** [comment]
```
Typeface.create(typeface, weight, italic);
```

This is only available to Android API >= 28.

**Comment by fabriziobertoglio1987** [comment]
> ```
> Typeface.create(typeface, weight, italic);
> ```
> 
> This is only available to Android API >= 28.

@stackia thanks a lot. You are right. I will try to fix it as soon as possible. Sorry. I wish you a good day. Fabrizio

**Comment by fabriziobertoglio1987** [comment]
@stackia Thanks a lot for your help

This is the screenshot of the functionality disabled on API < 28

**Comment by fabriziobertoglio1987** [comment]
relative to information included in commit https://github.com/facebook/react-native/pull/29117/commits/7dbeee527deecd3e41bfe91049e7eccec65de780 message. Log messages will notify the user of this issues.

As you can see below, rubik bold is not bold as they set `{fontWeight: 700}`, more information included in commit https://github.com/facebook/react-native/pull/29117/commits/7dbeee527deecd3e41bfe91049e7eccec65de780 message.

| **AFTER** | **AFTER** | 
|:-------------------------:|:-------------------------:|
|  |  |

| **AFTER** | **AFTER** | 
|:-------------------------:|:-------------------------:|
|  |  |

**Comment by mac-chong** [comment]
Hi, any idea when this can be deployed? We'd love to get our styling working with more recent version of React Native! :)

**Comment by SurajMDurgad** [comment]
Can anyone merge this branch? Why isn't this merged already? Styling is messed up in 0.63.2!

**Comment by dngconsulting** [comment]
Please, is it possible to merge this PR in the next version ?

**Comment by fangasvsass** [comment]
any update??

**Comment by fabriziobertoglio1987** [comment]
The only way to have this branch merged is testing it following instructions at https://github.com/facebook/react-native/wiki/Building-from-source#publish-your-own-version-of-react-native, react-native is growing in popularity and there has been a great increase in pull requests (300 open prs) so not all can be reviewed/tested by the Facebook team. I also prepared a video introduction trying to share all the knowledge earned while contributing to ReactNative, check it out at https://youtu.be/9RQ2GiApkzU to have a soft introduction to this pull request. I talk about this pull request in the video. Thanks a lot for the support. I wish you a good day!

**Comment by dngconsulting** [comment]
@fabriziobertoglio1987 Thanx for your answer (I got a look on your video) but still don't understand why this PR is not merged. You say :

> The only way to have this branch merged is testing it following instructions

Why this process has not been followed, who should do that ? The facebook team, the PR owner ? 

"

**Comment by FrederickEngelhardt** [comment]
Tried using https://www.npmjs.com/package/patch-package to bring in these changes to RN 0.62. Did not see any changes to the fonts for my android simulator API 29. Is patch-package not able to resolve this?

**Comment by fabriziobertoglio1987** [comment]
@FrederickEngelhardt you can follow this instructions  https://github.com/facebook/react-native/wiki/Building-from-source#publish-your-own-version-of-react-native to build the ReactAndroid binary with docker and you can configure your react native android project to use such version. Thanks

**Comment by LimanASL** [comment]
Any update about the merging?

File: ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTypefaceUtils.java [hunk 1]
```diff
@@ -81,34 +84,46 @@ public static Typeface applyStyles(
...
     }
-
-    if (typeface != null) {
-      return typeface;
-    } else {
-      return Typeface.defaultFromStyle(want);
+    if (numericBold || numericNormal) {
+      newStyle = numericBold ? Typeface.BOLD : Typeface.NORMAL;
+      typeface = Typeface.create(typeface, newStyle);
+      FLog.w(TAG, "Support for numeric font weight numeric values available only from Android API 28 Pie. Android device lower then API 28 will use normal or bold.");
```
Comment (fabriziobertoglio1987) [inline ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTypefaceUtils.java:124]: I will change the logs to `FLog.d` as seems more appropriate for this use case

**Comment by fabriziobertoglio1987** [comment]
>Please, is it possible to merge this PR in the next version ?

To answer the question of when this pr will be released in the next version of React Native, I suggest you to follow the discussions at https://github.com/react-native-community/releases/issues by subscribing to those threads. The process of releasing a new version with a bug fix is the following:

- an Issue is opened in facebook/react-native
- a contributor writes a Pull Request to fix/close the react-native Issue
- once the Pull Request is merged, the issue is closed. The review and merge is performed by the Facebook Team that works on the Facebook Ads App running React Native from source. Currently we have 351 open pull requests so the review and merge may take some time.
- The fix is tested in Facebook Ads App running React Native from master for several months
- For each realease (the upcoming one is 0.64.1) there will be a separate discussion on https://github.com/react-native-community/releases/issues to discuss
In this discussion everyone can request a commit to be cherrypicked from master. The link need to be from the a **commit** from master (for ex. https://github.com/facebook/react-native/pull/29157#issuecomment-651430945) and not a Pull Request not yet merged in master :peace_symbol: :pray:

**Comment by aliraza-noon** [comment]
can we get this reviewed and merged ? 

-- Thank you!

**Comment by MauriceArikoglu** [comment]
@dulmandakh 
Hi. You are a core contributor and were the author of #25341 - could you please review this PR and get this merged?

It's been 662 days since #25341's breaking change was introduced in [59.10](https://github.com/facebook/react-native/releases/tag/v0.59.10).

It's been 317 days since @fabriziobertoglio1987 provided a working fix.

Is there anything we can help you with to get this approved asap?

Best,
Maurice

**Comment by AlphaJuliettOmega** [comment]
It's bizarre to have only 'normal' and 'bold' font weights on Android but variable font sizes work on iOs
RN .64

on Android 700 defaults to 'Bold'
800+ becomes 'normal'

Now I have to implement platform specific font weights &/ somehow generate a font for each weight from a single font or disregard app designs to get the app to look similar on iOs + Android

**Comment by Kylar13** [comment]
Is there an update on this? How could we help get this through the door?

**Comment by yungsters** [comment]
Thank you for your contribution, @fabriziobertoglio1987. I just verified this locally (and made some notes for what this change means for Fabric) and expect it to be merged after Facebook CI succeeds.

File: ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTypefaceUtils.java (hunk 1 above)
Comment (yungsters) [reply ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTypefaceUtils.java:124]: I went ahead and made this change for you (in addition to formatting the file to conform to our lint rules). The merged commit should include this without any additional work from you. Thanks again!

Comment (fabriziobertoglio1987) [reply ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTypefaceUtils.java:124]: Thanks a lot @yungsters and sorry for not following up on this. I forgot about this comment. :pray:

**Comment by andreialecu** [comment]
If you'd like to help give this a better chance to be part of 0.64.3, upvote this comment for visibility in the upcoming v0.64.3 cherrypicks discussion:

https://github.com/react-native-community/releases/issues/232#issuecomment-860787470

**Comment by tushar-singhal** [comment]
this will be the part of 0.66 🎉

**Comment by manuhook** [comment]
> this will be the part of 0.66 🎉

Seems it's already in 0.65.x

**Comment by MauriceArikoglu** [comment]
@fabriziobertoglio1987 thanks for caring 👍

**Comment by pke** [comment]
This still doesn't work. I wonder if no one at facebook is using bold font faces in their apps?

**Comment by fabriziobertoglio1987** [comment]
https://github.com/facebook/react-native/issues/26193#issuecomment-926383825

**Pull request closed without merging**